{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OrthoEvolution_settings.txt\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "import glob\n",
    "import pandas as pd\n",
    "from Bio import Phylo\n",
    "from Bio import SeqIO\n",
    "import ete3\n",
    "from ete3 import Tree\n",
    "import os.path as path\n",
    "from datetime import date\n",
    "\n",
    "\n",
    "### First we need to get the information from a \"Settings_File\". There we have all the paths of all the files we are going to need. \n",
    "Settings_File = input()\n",
    "File_paths=[]\n",
    "if path.isfile(Settings_File) == True:\n",
    "    with open (Settings_File, 'r') as file:\n",
    "        for line in file.readlines():\n",
    "            File_paths.append(line.rstrip().split(':')[1])\n",
    "        file.close()\n",
    "       \n",
    "    #Name of the reference specie we are going to use\n",
    "    Specie_Name = File_paths[1]\n",
    "    Reference_Specie = File_paths[2]\n",
    "    #File called Duplications.tsv\n",
    "    Duplications_File = File_paths[3]+'/Gene_Duplication_Events/Duplications.tsv'\n",
    "    #Species labeled and non labeled trees\n",
    "    Species_Tree = File_paths[3]+'/Species_Tree/SpeciesTree_rooted_node_labels.txt'\n",
    "    Non_labeled_tree = File_paths[3]+'/Species_Tree/SpeciesTree_rooted.txt'\n",
    "    #Folder containing all the files of the Orthologues of the reference specie\n",
    "    Orthologues_Folder = File_paths[3]+'/Orthologues/Orthologues_' + Reference_Specie + '/' + Reference_Specie + '*.tsv'\n",
    "    #All the proteins of the reference specie used\n",
    "    Proteins_Fasta = File_paths[4]\n",
    "    #Files containing the orthogroups (Orthogroups.tsv) and the orthogroups of the Unassigned genes(Orthogroups_UnassignedGenes.tsv)\n",
    "    Orthogroups_File = File_paths[3] + '/Orthogroups/Orthogroups.tsv'\n",
    "    Orthogroup_UnassignedGenes = File_paths[3] + '/Orthogroups/Orthogroups_UnassignedGenes.tsv'\n",
    "    Output = File_paths[5]\n",
    "    Support = float(File_paths[6])\n",
    "    Added = ''\n",
    "    for i in range(0,len(Reference_Specie)):\n",
    "        if i >= len(Specie_Name):\n",
    "            Added += Reference_Specie[i]\n",
    "else:\n",
    "    Configurations = ['If you are not sure about how to use this settings.txt file: please read the README.md','Specie_Name:','Reference_Specie:','OrthoFinder_path:','Reference_Specie_fasta_path:','Output:','Support:']\n",
    "    with open ('OrthoEvolution_settings.txt','w') as file:\n",
    "        file.writelines(\"%s\\n\" % i for i in Configurations)\n",
    "    print('settings.txt file created')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------------------------------------------------------------------------------------------#\n",
    "#DUPLICATIONS\n",
    "#-----------------------------------------------------------------------------------------------#\n",
    "\n",
    "#DataFrame showing all the duplications provided by OrthoFinder.\n",
    "DUPLICATIONS_DF = pd.read_csv(Duplications_File, delimiter = \"\\t\", header=0)\n",
    "\n",
    "###############################################################################################################################################################\n",
    "#Functions:\n",
    "#With this function I am trying to get rid of all those duplicated genes in a Clade. It also renames the gene names in orden to have them all in the same format.\n",
    "def extract_duplication_events (Clade_duplication_table):\n",
    "    Total_dupl_genes = []\n",
    "    Genes_1 = []\n",
    "    Genes_2 = []\n",
    "    for gene in Clade_duplication_table['Genes 1']:\n",
    "        Genes_1.append(gene)\n",
    "    for gene in Clade_duplication_table['Genes 2']:\n",
    "        Genes_2.append(gene)\n",
    "    Genes_1 = ','.join(Genes_1)\n",
    "    Genes_1 = Genes_1.replace(' ','')\n",
    "    Genes_2 = ','.join(Genes_2)\n",
    "    Genes_2 = Genes_2.replace(' ','')\n",
    "    Unique_Genes = Genes_1.split(',') + Genes_2.split(',')\n",
    "    for gene in Unique_Genes:\n",
    "        if Reference_Specie in gene:\n",
    "            gene = gene.replace(Reference_Specie, '')\n",
    "            gene = gene[1:]\n",
    "            if gene not in Total_dupl_genes:\n",
    "                Total_dupl_genes.append(gene)\n",
    "    return Total_dupl_genes\n",
    "\n",
    "#This function is used to obtain the species names from the oldest one to our Reference_Specie\n",
    "def get_subtree (Reference_Specie, Tree):\n",
    "    Clades = []\n",
    "    for i in range(0,2):\n",
    "        if Reference_Specie not in Tree[i].get_leaf_names():\n",
    "            Clades.append(Tree[i].get_leaf_names())\n",
    "        else:\n",
    "            Subtree = Tree[i].get_children()\n",
    "    return Clades, Subtree\n",
    "\n",
    "#The following function is used to shorten the names of the species so it is easier to read them.\n",
    "#Función para dejar los nombres de las especies sin adornos -> Cambiar según los nombres que le hayas puesto a las especies, se pueden añadir o quitar ifs y si se usa la misma estructura que tienen estos no debería haber problema.\n",
    "def short_names (Species, Added):\n",
    "    Clades_short = []\n",
    "    for species in Species:\n",
    "        if Added in species:\n",
    "            Clades_short.append(species.replace(Added, ''))\n",
    "    return Clades_short"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "#################################################################################################################################################################\n",
    "#We are going to get the clades we are interested in (related to the reference specie) directly from the Species_Tree\n",
    "S_tree = Phylo.read(Species_Tree, \"newick\")\n",
    "\n",
    "Tree_Path = []\n",
    "for clade in S_tree.trace('N0', Reference_Specie):\n",
    "    c = str(clade)\n",
    "    Tree_Path.append(c)\n",
    "Tree_Path = list(reversed(Tree_Path))\n",
    "Tree_Path.remove(Reference_Specie)\n",
    "#Para poder buscar las Clades en el árbol necesitas que Tree_Path no contenga la especie referencia. Se vuelve a meter después.\n",
    "\n",
    "#Names from the species tree -> With this we obtain a list of lists (if some species are agrupated they will be inside a list)\n",
    "Tree = ete3.Tree(Non_labeled_tree)\n",
    "Clades = []\n",
    "for i in range(0,len(Tree_Path)):\n",
    "    if i == 0:\n",
    "        Clades += get_subtree(Reference_Specie, Tree.get_children())[0]\n",
    "        Subtree = get_subtree(Reference_Specie, Tree.get_children())[1]\n",
    "    else:\n",
    "        Clades += get_subtree(Reference_Specie, Subtree)[0]\n",
    "        Subtree = get_subtree(Reference_Specie, Subtree)[1]\n",
    "Clades = list(reversed(Clades))\n",
    "    \n",
    "#It is useful though to have all the species names in just one list:        \n",
    "Clades_Only = []\n",
    "for element in Clades:\n",
    "    if len(element) == 1:\n",
    "        Clades_Only += element\n",
    "    else:\n",
    "        for clade in element:\n",
    "            Clades_Only.append(clade)\n",
    "Clades_Only = short_names(Clades_Only, Added)\n",
    "    \n",
    "Species_Path = []\n",
    "for element in Clades:\n",
    "    if len(element) == 1:\n",
    "        Species_Path += short_names(element, Added)\n",
    "            \n",
    "    if len(element) > 1:\n",
    "        Species_Path.append(element[1].replace(Added,'') + '_group')\n",
    "            \n",
    "\n",
    "#This code is used to now the node in wich the novo gene is originated\n",
    "#Poner en el diccionario el easy_node\n",
    "Clade_Path = {Specie_Name:Specie_Name}\n",
    "for i in range(0,len(Tree_Path)):\n",
    "    Clade_Path[Species_Path[i]] = Tree_Path[i]\n",
    "    \n",
    "#Para los siguientes pasos necesitas que la especie referencia esté en Tree_Path, para que salga ordenado tienes que girar la lista.\n",
    "Tree_Path = list(reversed(Tree_Path))\n",
    "Tree_Path.append(Reference_Specie)\n",
    "Tree_Path = list(reversed(Tree_Path))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "#I am obtaining the duplicated genes from each Clade (all repetitions inside each Clade are already eliminated)\n",
    "Duplications_by_Clade_dic = {}\n",
    "for i in range(0,len(Tree_Path)):\n",
    "    One_Clade = []\n",
    "    duplication_events_table = DUPLICATIONS_DF[(DUPLICATIONS_DF['Species Tree Node'] == Tree_Path[i]) & (DUPLICATIONS_DF['Support'] >= Support)]\n",
    "    One_Clade += extract_duplication_events(duplication_events_table)\n",
    "    Duplications_by_Clade_dic[Tree_Path[i]] = One_Clade\n",
    "\n",
    "\n",
    "\n",
    "#Now I am comparing each Clade with each other in order to eliminate in between clades repetitions. As it is sorted from newer duplications to older the newest duplicated genes will be kept and the older one will be removed\n",
    "All_Copies_dic = {}\n",
    "for i in range(0,len(Tree_Path)):\n",
    "    Copies_dic = {}\n",
    "    for j in range(1, len(Tree_Path)):\n",
    "        Copies_per_clade = []\n",
    "        if Tree_Path[i] == Tree_Path[j]:\n",
    "            continue\n",
    "        elif j < i:\n",
    "            continue\n",
    "        else:\n",
    "            for element in Duplications_by_Clade_dic.get(Tree_Path[j]):\n",
    "                if element in Duplications_by_Clade_dic.get(Tree_Path[i]):\n",
    "                    Copies_per_clade.append(element)\n",
    "        Copies_dic[Tree_Path[j]] = Copies_per_clade\n",
    "    All_Copies_dic[Tree_Path[i]] = Copies_dic\n",
    "\n",
    "    \n",
    "for i in range(0, len(Tree_Path)):\n",
    "    for j in range(1, len(Tree_Path)):\n",
    "        if Tree_Path[i] == Tree_Path[j]:\n",
    "            continue\n",
    "        elif j < i:\n",
    "            continue\n",
    "        else:\n",
    "            for element in All_Copies_dic.get(Tree_Path[i]).get(Tree_Path[j]):  \n",
    "                if element in Duplications_by_Clade_dic.get(Tree_Path[j]):\n",
    "                    Duplications_by_Clade_dic.get(Tree_Path[j]).remove(element)\n",
    "        \n",
    "\n",
    "#Here we are getting the duplication events in each node. \n",
    "Duplication_events = []\n",
    "for i in range(0,len(Tree_Path)):\n",
    "    table = DUPLICATIONS_DF[(DUPLICATIONS_DF['Species Tree Node'] == Tree_Path[i]) & (DUPLICATIONS_DF['Support'] >= Support)]\n",
    "    Duplication_events.append(len(table))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "#It is important to be certain about the duplication events of the reference specie, that is why we are calculating that number in a diferent way:\n",
    "Reference_Specie_DF = DUPLICATIONS_DF[(DUPLICATIONS_DF['Species Tree Node'] == Reference_Specie) & (DUPLICATIONS_DF['Support'] >= Support)] \n",
    "\n",
    "Orthogroups_copies = []\n",
    "for Orthogroup in Reference_Specie_DF['Orthogroup']:\n",
    "    Orthogroups_copies.append(Orthogroup)\n",
    "\n",
    "Orthogroups = []    \n",
    "for Orthogroup in Orthogroups_copies:\n",
    "    if Orthogroup not in Orthogroups:\n",
    "        Orthogroups.append(Orthogroup)\n",
    "\n",
    "Total_genes_Orthogroup = []\n",
    "for Orthogroup in Orthogroups:\n",
    "    Genes_Orthogroup = []\n",
    "    ORTHOGROUP_DF = Reference_Specie_DF[(Reference_Specie_DF['Orthogroup'] == Orthogroup)]\n",
    "    for i in range(0,len(ORTHOGROUP_DF)):\n",
    "        Genes_Orthogroup += ((ORTHOGROUP_DF.iloc[i]['Genes 1'].replace(' ','') + ',' + ORTHOGROUP_DF.iloc[i]['Genes 2'].replace(' ','')).split(','))\n",
    "    Total_genes_Orthogroup.append(Genes_Orthogroup) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "Non_duplicated_genes_Orthogroup=[]\n",
    "for element in Total_genes_Orthogroup:\n",
    "    Non_duplicated_genes_Orthogroup.append(list(set(element)))\n",
    "\n",
    "Duplication_Events_Ref_Spe = 0\n",
    "for element in Non_duplicated_genes_Orthogroup:\n",
    "    Duplication_Events_Ref_Spe += (len(element)-1)        \n",
    "\n",
    "#In order to normalize the duplication events we need to now the branch length of each node.\n",
    "Branch_Length = {}\n",
    "for node in S_tree.find_clades(branch_length = True):\n",
    "    Branch_Length[node.name] = node.branch_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################################################################################################################################################################################    \n",
    "Duplications_file = open(str(date.today()) + '-' + Output + \"_Duplication_Analysis.tsv\", \"w\")\n",
    "\n",
    "writer = csv.writer(Duplications_file, delimiter = '\\t')\n",
    "writer.writerow(['Node','Genes'])\n",
    "for key, value in Duplications_by_Clade_dic.items():\n",
    "    for gene in value:\n",
    "        writer.writerow([key.replace(Added, ''), gene])\n",
    "\n",
    "Duplications_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis done, new file: 2020-12-04-pruebas_Duplication_Analysis.tsv\n",
      "Analysis done, new file: 2020-12-04-pruebas_Duplication_Overall.tsv\n"
     ]
    }
   ],
   "source": [
    "##########################################################################################################################################################################################\n",
    "#Creating a tsv file to save the overall information of the Gene Duplications Analysis\n",
    "Duplications_Overall = open(str(date.today()) + '-' + Output + \"_Duplication_Overall.tsv\", \"w\")\n",
    "\n",
    "writer = csv.writer(Duplications_Overall, delimiter = '\\t')\n",
    "\n",
    "counter = 0\n",
    "writer.writerow(['Node','Specie','Duplications', 'Dupl_Events', 'Branch_length', 'Normalized_events'])\n",
    "for key, value in Duplications_by_Clade_dic.items():\n",
    "    if Specie_Name in key.replace(Added, ''):\n",
    "        writer.writerow([key.replace(Added, ''), list(Clade_Path.keys())[counter], len(value),Duplication_Events_Ref_Spe, Branch_Length.get(key),(Duplication_Events_Ref_Spe)/(Branch_Length.get(key)*1000)])  \n",
    "    if key.replace(Added, '') not in Branch_Length.keys() and Specie_Name not in key.replace(Added, ''):\n",
    "        writer.writerow([key.replace(Added, ''), list(Clade_Path.keys())[counter],len(value),Duplication_events[counter],0,0])\n",
    "    elif Specie_Name not in key:\n",
    "        writer.writerow([key.replace(Added, ''), list(Clade_Path.keys())[counter],len(value),Duplication_events[counter], Branch_Length.get(key.replace(Added, '')),(Duplication_events[counter])/(Branch_Length.get(key.replace(Added, ''))*1000)])\n",
    "\n",
    "    counter += 1\n",
    "\n",
    "Duplications_Overall.close()\n",
    "\n",
    "print(\"Analysis done, new file: \" + str(date.today()) + '-' + Output + \"_Duplication_Analysis.tsv\")\n",
    "print(\"Analysis done, new file: \" + str(date.today()) + '-' + Output + \"_Duplication_Overall.tsv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Drosophila_melanogaster'"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Specie_Name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
