{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/mhuertas/GeneBPhylo/GeneBPhylo_settings.txt\n",
      "Analysis done, new file: 2021-07-05-Dmel_Duplication_Analysis.tsv\n",
      "Analysis done, new file: 2021-07-05-Dmel_Duplication_Overall.tsv\n",
      "Analysis done, new file:  2021-07-05-Dmel_De_novo_Analysis.tsv\n",
      "Analysis done, new file:  2021-07-05-Dmel_De_novo_Overall.tsv\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "import glob\n",
    "import pandas as pd\n",
    "from Bio import Phylo\n",
    "from Bio import SeqIO\n",
    "import ete3\n",
    "from ete3 import Tree\n",
    "import os.path as path\n",
    "from datetime import date\n",
    "\n",
    "\n",
    "### First we need to get the information from a \"Settings_File\". There we have all the paths of all the files we are going to need. \n",
    "Settings_File = input()\n",
    "File_paths=[]\n",
    "if path.isfile(Settings_File) == True:\n",
    "    with open (Settings_File, 'r') as file:\n",
    "        for line in file.readlines():\n",
    "            File_paths.append(line.rstrip().split(':')[1])\n",
    "        file.close()\n",
    "       \n",
    "    #Name of the reference specie we are going to use\n",
    "    Specie_Name = File_paths[1]\n",
    "    Reference_Specie = File_paths[2]\n",
    "    #File called Duplications.tsv\n",
    "    Duplications_File = File_paths[3]+'/Gene_Duplication_Events/Duplications.tsv'\n",
    "    #Species labeled and non labeled trees\n",
    "    Species_Tree = File_paths[3]+'/Species_Tree/SpeciesTree_rooted_node_labels.txt'\n",
    "    Non_labeled_tree = File_paths[3]+'/Species_Tree/SpeciesTree_rooted.txt'\n",
    "    #Folder containing all the files of the Orthologues of the reference specie\n",
    "    Orthologues_Folder = File_paths[3]+'/Orthologues/Orthologues_' + Reference_Specie + '/' + Reference_Specie + '*.tsv'\n",
    "    #All the proteins of the reference specie used\n",
    "    Proteins_Fasta = File_paths[4]\n",
    "    #Files containing the orthogroups (Orthogroups.tsv) and the orthogroups of the Unassigned genes(Orthogroups_UnassignedGenes.tsv)\n",
    "    Orthogroups_File = File_paths[3] + '/Orthogroups/Orthogroups.tsv'\n",
    "    Orthogroup_UnassignedGenes = File_paths[3] + '/Orthogroups/Orthogroups_UnassignedGenes.tsv'\n",
    "    Output = File_paths[5]\n",
    "    Support = float(File_paths[6])\n",
    "    Added = ''\n",
    "    for i in range(0,len(Reference_Specie)):\n",
    "        if i >= len(Specie_Name):\n",
    "            Added += Reference_Specie[i]\n",
    "    \n",
    "\n",
    "    #-----------------------------------------------------------------------------------------------#\n",
    "    #DUPLICATIONS\n",
    "    #-----------------------------------------------------------------------------------------------#\n",
    "\n",
    "    #DataFrame showing all the duplications provided by OrthoFinder.\n",
    "    DUPLICATIONS_DF = pd.read_csv(Duplications_File, delimiter = \"\\t\", header=0, low_memory=False)\n",
    "    for c in DUPLICATIONS_DF.columns:\n",
    "        DUPLICATIONS_DF[c] = DUPLICATIONS_DF[c].astype(\"object\")\n",
    "    ###############################################################################################################################################################\n",
    "    #Functions:\n",
    "    #With this function I am trying to get rid of all those duplicated genes in a Clade. It also renames the gene names in orden to have them all in the same format.\n",
    "    def extract_duplication_events (Clade_duplication_table):\n",
    "        Total_dupl_genes = []\n",
    "        Genes_1 = []\n",
    "        Genes_2 = []\n",
    "        for gene in Clade_duplication_table['Genes 1']:\n",
    "            Genes_1.append(gene)\n",
    "        for gene in Clade_duplication_table['Genes 2']:\n",
    "            Genes_2.append(gene)\n",
    "        Genes_1 = ','.join(Genes_1)\n",
    "        Genes_1 = Genes_1.replace(' ','')\n",
    "        Genes_2 = ','.join(Genes_2)\n",
    "        Genes_2 = Genes_2.replace(' ','')\n",
    "        Unique_Genes = Genes_1.split(',') + Genes_2.split(',')\n",
    "        for gene in Unique_Genes:\n",
    "            if Reference_Specie in gene:\n",
    "                gene = gene.replace(Reference_Specie, '')\n",
    "                gene = gene[1:]\n",
    "                if gene not in Total_dupl_genes:\n",
    "                    Total_dupl_genes.append(gene)\n",
    "        return Total_dupl_genes\n",
    "\n",
    "    #This function is used to obtain the species names from the oldest one to our Reference_Specie\n",
    "    def get_subtree (Reference_Specie, Tree):\n",
    "        Clades = []\n",
    "        for i in range(0,2):\n",
    "            if Reference_Specie not in Tree[i].get_leaf_names():\n",
    "                Clades.append(Tree[i].get_leaf_names())\n",
    "            else:\n",
    "                Subtree = Tree[i].get_children()\n",
    "        return Clades, Subtree\n",
    "\n",
    "    #The following function is used to shorten the names of the species so it is easier to read them.\n",
    "    #Función para dejar los nombres de las especies sin adornos -> Cambiar según los nombres que le hayas puesto a las especies, se pueden añadir o quitar ifs y si se usa la misma estructura que tienen estos no debería haber problema.\n",
    "    def short_names (Species, Added):\n",
    "        Clades_short = []\n",
    "        for species in Species:\n",
    "            if Added in species:\n",
    "                Clades_short.append(species.replace(Added, ''))\n",
    "        return Clades_short\n",
    "\n",
    "    #################################################################################################################################################################\n",
    "    #We are going to get the clades we are interested in (related to the reference specie) directly from the Species_Tree\n",
    "    S_tree = Phylo.read(Species_Tree, \"newick\")\n",
    "\n",
    "    Tree_Path = []\n",
    "    for clade in S_tree.trace('N0', Reference_Specie):\n",
    "        c = str(clade)\n",
    "        Tree_Path.append(c)\n",
    "    Tree_Path = list(reversed(Tree_Path))\n",
    "    Tree_Path.remove(Reference_Specie)\n",
    "    #Para poder buscar las Clades en el árbol necesitas que Tree_Path no contenga la especie referencia. Se vuelve a meter después.\n",
    "\n",
    "    #Names from the species tree -> With this we obtain a list of lists (if some species are agrupated they will be inside a list)\n",
    "    Tree = ete3.Tree(Non_labeled_tree)\n",
    "    Clades = []\n",
    "    for i in range(0,len(Tree_Path)):\n",
    "        if i == 0:\n",
    "            Clades += get_subtree(Reference_Specie, Tree.get_children())[0]\n",
    "            Subtree = get_subtree(Reference_Specie, Tree.get_children())[1]\n",
    "        else:\n",
    "            Clades += get_subtree(Reference_Specie, Subtree)[0]\n",
    "            Subtree = get_subtree(Reference_Specie, Subtree)[1]\n",
    "    Clades = list(reversed(Clades))\n",
    "    \n",
    "    #It is useful though to have all the species names in just one list:        \n",
    "    Clades_Only = []\n",
    "    for element in Clades:\n",
    "        if len(element) == 1:\n",
    "            Clades_Only += element\n",
    "        else:\n",
    "            for clade in element:\n",
    "                Clades_Only.append(clade)\n",
    "    Clades_Only = short_names(Clades_Only, Added)\n",
    "    \n",
    "    Species_Path = []\n",
    "    for element in Clades:\n",
    "        if len(element) == 1:\n",
    "            Species_Path += short_names(element, Added)\n",
    "            \n",
    "        if len(element) > 1:\n",
    "            Species_Path.append(element[1].replace(Added,'') + '_group')\n",
    "            \n",
    "\n",
    "    #This code is used to now the node in wich the novo gene is originated\n",
    "    #Poner en el diccionario el easy_node\n",
    "    Clade_Path = {Specie_Name:Specie_Name}\n",
    "    for i in range(0,len(Tree_Path)):\n",
    "        Clade_Path[Species_Path[i]] = Tree_Path[i]\n",
    "    \n",
    "    #Para los siguientes pasos necesitas que la especie referencia esté en Tree_Path, para que salga ordenado tienes que girar la lista.\n",
    "    Tree_Path = list(reversed(Tree_Path))\n",
    "    Tree_Path.append(Reference_Specie)\n",
    "    Tree_Path = list(reversed(Tree_Path))\n",
    "\n",
    "    #I am obtaining the duplicated genes from each Clade (all repetitions inside each Clade are already eliminated)\n",
    "    #I am obtaining the duplicated genes from each Clade (all repetitions inside each Clade are already eliminated)\n",
    "    Duplications_by_Clade_dic = {}\n",
    "    for i in range(0,len(Tree_Path)):\n",
    "        One_Clade = []\n",
    "        duplication_events_table = DUPLICATIONS_DF[(DUPLICATIONS_DF['Species Tree Node'] == Tree_Path[i]) & (DUPLICATIONS_DF['Support'] >= Support)]\n",
    "        One_Clade += extract_duplication_events(duplication_events_table)\n",
    "        Duplications_by_Clade_dic[Tree_Path[i]] = One_Clade\n",
    "\n",
    "    #Now I am comparing each Clade with each other in order to eliminate in between clades repetitions. As it is sorted from newer duplications to older the newest duplicated genes will be kept and the older one will be removed\n",
    "    All_Copies_dic = {}\n",
    "    for i in range(0,len(Tree_Path)):\n",
    "        Copies_dic = {}\n",
    "        for j in range(1, len(Tree_Path)):\n",
    "            Copies_per_clade = []\n",
    "            if Tree_Path[i] == Tree_Path[j]:\n",
    "                continue\n",
    "            elif j < i:\n",
    "                continue\n",
    "            else:\n",
    "                for element in Duplications_by_Clade_dic.get(Tree_Path[j]):\n",
    "                    if element in Duplications_by_Clade_dic.get(Tree_Path[i]):\n",
    "                        Copies_per_clade.append(element)\n",
    "            Copies_dic[Tree_Path[j]] = Copies_per_clade\n",
    "        All_Copies_dic[Tree_Path[i]] = Copies_dic\n",
    "\n",
    "\n",
    "    for i in range(0, len(Tree_Path)):\n",
    "        for j in range(1, len(Tree_Path)):\n",
    "            if Tree_Path[i] == Tree_Path[j]:\n",
    "                continue\n",
    "            elif j < i:\n",
    "                continue\n",
    "            else:\n",
    "                for element in All_Copies_dic.get(Tree_Path[i]).get(Tree_Path[j]):  \n",
    "                    if element in Duplications_by_Clade_dic.get(Tree_Path[j]):\n",
    "                        Duplications_by_Clade_dic.get(Tree_Path[j]).remove(element)\n",
    "\n",
    "\n",
    "    #Here we are getting the duplication events in each node. \n",
    "    Duplication_events = []\n",
    "    for i in range(0,len(Tree_Path)):\n",
    "        table = DUPLICATIONS_DF[(DUPLICATIONS_DF['Species Tree Node'] == Tree_Path[i]) & (DUPLICATIONS_DF['Support'] >= Support)]\n",
    "        Duplication_events.append(len(table))\n",
    "\n",
    "    #It is important to be certain about the duplication events of the reference specie, that is why we are calculating that number in a diferent way:\n",
    "    Reference_Specie_DF = DUPLICATIONS_DF[(DUPLICATIONS_DF['Species Tree Node'] == Reference_Specie) & (DUPLICATIONS_DF['Support'] >= Support)] \n",
    "\n",
    "    Orthogroups_copies = []\n",
    "    for Orthogroup in Reference_Specie_DF['Orthogroup']:\n",
    "        Orthogroups_copies.append(Orthogroup)\n",
    "\n",
    "    Orthogroups = []    \n",
    "    for Orthogroup in Orthogroups_copies:\n",
    "        if Orthogroup not in Orthogroups:\n",
    "            Orthogroups.append(Orthogroup)\n",
    "\n",
    "    Total_genes_Orthogroup = []\n",
    "    for Orthogroup in Orthogroups:\n",
    "        Genes_Orthogroup = []\n",
    "        ORTHOGROUP_DF = Reference_Specie_DF[(Reference_Specie_DF['Orthogroup'] == Orthogroup)]\n",
    "        for i in range(0,len(ORTHOGROUP_DF)):\n",
    "            Genes_Orthogroup += ((ORTHOGROUP_DF.iloc[i]['Genes 1'].replace(' ','') + ',' + ORTHOGROUP_DF.iloc[i]['Genes 2'].replace(' ','')).split(','))\n",
    "        Total_genes_Orthogroup.append(Genes_Orthogroup) \n",
    "\n",
    "    Non_duplicated_genes_Orthogroup=[]\n",
    "    for element in Total_genes_Orthogroup:\n",
    "        Non_duplicated_genes_Orthogroup.append(list(set(element)))\n",
    "\n",
    "    Duplication_Events_Ref_Spe = 0\n",
    "    for element in Non_duplicated_genes_Orthogroup:\n",
    "        Duplication_Events_Ref_Spe += (len(element)-1)        \n",
    "\n",
    "    #In order to normalize the duplication events we need to now the branch length of each node.\n",
    "    Branch_Length = {}\n",
    "    for node in S_tree.find_clades(branch_length = True):\n",
    "        Branch_Length[node.name] = node.branch_length\n",
    "\n",
    "    ##########################################################################################################################################################################################    \n",
    "    #Creating a tsv file to save all the duplications of each node\n",
    "    ##########################################################################################################################################################################################    \n",
    "    Duplications_file = open(str(date.today()) + '-' + Output + \"_Duplication_Analysis.tsv\", \"w\")\n",
    "\n",
    "    writer = csv.writer(Duplications_file, delimiter = '\\t')\n",
    "    writer.writerow(['Node','Genes'])\n",
    "    for key, value in Duplications_by_Clade_dic.items():\n",
    "        for gene in value:\n",
    "            writer.writerow([key.replace(Added, ''), gene])\n",
    "\n",
    "    Duplications_file.close()\n",
    "\n",
    "    ##########################################################################################################################################################################################\n",
    "    #Creating a tsv file to save the overall information of the Gene Duplications Analysis\n",
    "    ##########################################################################################################################################################################################\n",
    "    Duplications_Overall = open(str(date.today()) + '-' + Output + \"_Duplication_Overall.tsv\", \"w\")\n",
    "\n",
    "    writer = csv.writer(Duplications_Overall, delimiter = '\\t')\n",
    "\n",
    "    counter = 0\n",
    "    writer.writerow(['Node','Specie','Duplications', 'Dupl_Events', 'Branch_length', 'Normalized_events'])\n",
    "    for key, value in Duplications_by_Clade_dic.items():\n",
    "        if Specie_Name in key.replace(Added, ''):\n",
    "            writer.writerow([key.replace(Added, ''), list(Clade_Path.keys())[counter], len(value),Duplication_Events_Ref_Spe, Branch_Length.get(key),(Duplication_Events_Ref_Spe)/(Branch_Length.get(key)*1000)])  \n",
    "        if key.replace(Added, '') not in Branch_Length.keys() and Specie_Name not in key.replace(Added, ''):\n",
    "            writer.writerow([key.replace(Added, ''), list(Clade_Path.keys())[counter],len(value),Duplication_events[counter],0,0])\n",
    "        elif Specie_Name not in key:\n",
    "            writer.writerow([key.replace(Added, ''), list(Clade_Path.keys())[counter],len(value),Duplication_events[counter], Branch_Length.get(key.replace(Added, '')),(Duplication_events[counter])/(Branch_Length.get(key.replace(Added, ''))*1000)])\n",
    "\n",
    "        counter += 1\n",
    "\n",
    "    Duplications_Overall.close()\n",
    "\n",
    "    print(\"Analysis done, new file: \" + str(date.today()) + '-' + Output + \"_Duplication_Analysis.tsv\")\n",
    "    print(\"Analysis done, new file: \" + str(date.today()) + '-' + Output + \"_Duplication_Overall.tsv\")\n",
    "\n",
    "\n",
    "\n",
    "    #------------------------------------------------------------------------------------------------------------#\n",
    "    #DE NOVO\n",
    "    #------------------------------------------------------------------------------------------------------------#\n",
    "    ###########################################################################################################################################################################################\n",
    "    #Function:\n",
    "    #The following function is used to transform a DataFrame into a dictionary(key=Gene: value=Orthogroup)\n",
    "    #Te coge un dataframe y te lo transforma en un diccionario del tipo -> Gen:Orthogroup\n",
    "    def orthogroups_todict (DataFrame, Species):\n",
    "        Orthogroup_Specie = DataFrame[['Orthogroup',Species]].dropna()\n",
    "        Orthogroups_DICT = Orthogroup_Specie.set_index('Orthogroup')[Species].to_dict()\n",
    "        Orthogroup_Genes = {}\n",
    "        for element in Orthogroups_DICT:\n",
    "            Orthogroup_Genes[element] = Orthogroups_DICT[element].replace(' ','').split(',')\n",
    "        Gene_Orthogroup = {}\n",
    "        for element in Orthogroup_Genes:\n",
    "            for gene in Orthogroup_Genes.get(element):\n",
    "                Gene_Orthogroup[gene] = element\n",
    "        return Gene_Orthogroup\n",
    "\n",
    "    ##############################################################################################################################################################################################\n",
    "\n",
    "    #First we transform all the files in the Orthologues folder into DataFrames\n",
    "    #Obtener todos los archivos de Orthologos y guardarlos en una lista de DataFrames\n",
    "    Orthologues_Files = glob.glob(Orthologues_Folder)\n",
    "\n",
    "    Orthologues_Tables = []\n",
    "    for file in Orthologues_Files:\n",
    "        data = pd.read_csv(file, sep = '\\t', low_memory=False)\n",
    "        Orthologues_Tables.append(data)\n",
    "\n",
    "    #Here we are obtaining all the orthologues in the dataframe and saving them in lists.    \n",
    "    #Lista de listas. En cada lista hay los Ortologos de genes de la especie referencia de cada especie\n",
    "    Orthologues = []\n",
    "    for table in Orthologues_Tables:\n",
    "        Group_genes=[]\n",
    "        for gene in table[Reference_Specie]:\n",
    "            Group_genes.append(gene.replace(' ',''))\n",
    "        Total_genes = ','.join(Group_genes)\n",
    "        Group_genes = Total_genes.split(',')\n",
    "        Orthologues.append(Group_genes)\n",
    "\n",
    "    #We are going to get the names of the species from the Orthologues_Folder.\n",
    "    #Lista con el nombre de las especies (el orden sigue el orden de los DataFrames). Se usa porque está en el mismo orden que la lista de listas con los ortologos y así se hace el diccionario\n",
    "\n",
    "    #Names from the Orthologues_Folder\n",
    "    Species_Names = []\n",
    "    for table in Orthologues_Tables:\n",
    "        Species_Names.append(table.columns[2])\n",
    "    Species_Names = short_names(Species_Names, Added)\n",
    "\n",
    "    #Para los siguientes pasos necesitas que está lista no contenga la especie referencia\n",
    "    Tree_Path.remove(Reference_Specie)\n",
    "\n",
    "    #Here we are creating a dictionary in order to now wich Orthologues are from what specie. We are interested in getting the lists of Orthologues ordered according to the tree path\n",
    "    #Hacer diccionario con las especies, como el orden es el de las orthologues_tables, se usa Species_names y no Clades.\n",
    "    Species_Genes = {}\n",
    "    for i in range(0,len(Species_Names)):\n",
    "        Species_Genes[Species_Names[i]] = Orthologues[i]\n",
    "    #A partir del diccionario podemos tener una lista de los genes ordenados por especie. Esto se usa para saber que genes de la especie de referencia son únicos\n",
    "    Orthologues_ordered_list = []\n",
    "    for i in range(0,len(Clades_Only)):\n",
    "        Orthologues_ordered_list += Species_Genes.get(Clades_Only[i])\n",
    "\n",
    "    #Once we have all the orthologues we are interested in knowing wich ones are exclusive of the reference specie. First we have to get all the proteins used in the study.\n",
    "    #Obtener una lista de todas las proteinas de la especie de referencia utilizadas en el estudio\n",
    "    All_genes = []\n",
    "    for record in SeqIO.parse(Proteins_Fasta, \"fasta\"):\n",
    "        All_genes.append(str(record.description).replace(' ',''))\n",
    "\n",
    "    #Here we compare all the proteins used with those we have inside an Orthogroup and as a result we have the proteins exclusive of the reference specie.\n",
    "    #Lista en la que están las proteinas de novo en Scer. \n",
    "    ReferenceSpecie_de_novo = []\n",
    "    for genes in All_genes:\n",
    "        if genes not in Orthologues_ordered_list:\n",
    "            ReferenceSpecie_de_novo.append(genes)\n",
    "\n",
    "    #Once the reference specie is done we have to sort de novo genes by specie.   \n",
    "    #Lista con los genes de novo (no Especie de referencia). Cada lista dentro de la lista contiene los genes de una especie, ordenado según el árbol filogenético.\n",
    "    nonExclusive_de_novo = []\n",
    "    counter = 0\n",
    "    for i in range(0,len(Clades_Only)):\n",
    "        Species_de_novo = []\n",
    "        counter += len(Species_Genes.get(Clades_Only[i]))\n",
    "        for gene in Species_Genes.get(Clades_Only[i]):\n",
    "            if gene not in Orthologues_ordered_list[counter:]:\n",
    "                Species_de_novo.append(gene)\n",
    "        nonExclusive_de_novo.append(Species_de_novo) \n",
    "\n",
    "    #Diccionario --> Especie: Genes de novo\n",
    "    De_novo_genes = {}\n",
    "    De_novo_genes[Specie_Name] = ReferenceSpecie_de_novo\n",
    "    for i in range(0,len(Clades_Only)):\n",
    "        De_novo_genes[Clades_Only[i]] = nonExclusive_de_novo[i]\n",
    "\n",
    "    #Now we want to have every de novo protein idexed by specie and by Orthogroup. In order to do that we need all the Orthogroups provided by OrthoFinder\n",
    "    #Extraer los Ortogrupos de las proteínas de todas las especies\n",
    "    Orthogroups_non_exclusive = pd.read_csv(Orthogroups_File, delimiter = \"\\t\", header=0,low_memory=False)\n",
    "    for c in Orthogroups_non_exclusive.columns:\n",
    "        Orthogroups_non_exclusive[c] = Orthogroups_non_exclusive[c].astype(\"object\")\n",
    "    #Extraer Orthogrupos exclusivos de la especie de referencia.\n",
    "    Orthogroups_Exclusive = pd.read_csv(Orthogroup_UnassignedGenes, delimiter = \"\\t\", header=0, low_memory=False)\n",
    "    for c in Orthogroups_Exclusive.columns:\n",
    "        Orthogroups_Exclusive[c] = Orthogroups_Exclusive[c].astype(\"object\")\n",
    "\n",
    "    #Here we are using orthogroups_todict to get each protein related to an Orthogroup.\n",
    "    Orthogroups_non_exclusive_dict = orthogroups_todict(Orthogroups_non_exclusive, Reference_Specie)\n",
    "    Orthogroups_exclusive_dict = orthogroups_todict(Orthogroups_Exclusive, Reference_Specie)\n",
    "\n",
    "    #In order to have all the information of the protein in just one object we are going to create a dictionary (key = protein : values = [Orthogroup, Specie, Node])\n",
    "    De_novo_gene_information = {}\n",
    "    for specie in De_novo_genes:\n",
    "        for gene in De_novo_genes.get(specie):\n",
    "            if gene in list(Orthogroups_non_exclusive_dict.keys()):\n",
    "                De_novo_gene_information[gene] = [Orthogroups_non_exclusive_dict.get(gene), specie]\n",
    "            elif gene in list(Orthogroups_exclusive_dict.keys()):\n",
    "                De_novo_gene_information[gene] = [Orthogroups_exclusive_dict.get(gene), specie]\n",
    "\n",
    "#Las especies agrupadas pasan a llamarse todas con un único nombre.\n",
    "    #Si hay más de una agrupación se llamaran el nombre que queramos + un número (empezando con 1 la agrupación más cercana a la especie de referencia)\n",
    "    for element in Clades:\n",
    "        if len(element) > 1:\n",
    "            for specie in De_novo_gene_information:\n",
    "                if De_novo_gene_information[specie][1] in short_names(element, Added):\n",
    "                    De_novo_gene_information[specie][1] = element[1].replace(Added, '') + '_group'\n",
    "\n",
    "    for element in De_novo_gene_information:\n",
    "        De_novo_gene_information[element].append(Clade_Path[De_novo_gene_information[element][1]])\n",
    "\n",
    "    #In order to obtain the number of 'de novo' events we are going to count the number of Orthogroups    \n",
    "    Species_Path = list(reversed(Species_Path))\n",
    "    Species_Path.append(Specie_Name)\n",
    "    Species_Path = list(reversed(Species_Path))\n",
    "\n",
    "    count_Orthogroups = []\n",
    "    for i in range(0, len(Species_Path)):\n",
    "        Orthogroup = []\n",
    "        for gene in De_novo_gene_information:\n",
    "            if De_novo_gene_information[gene][1] == Species_Path[i]:\n",
    "                if De_novo_gene_information[gene][0] not in Orthogroup:\n",
    "                    Orthogroup.append(De_novo_gene_information[gene][0])\n",
    "        count_Orthogroups.append(len(Orthogroup))\n",
    "\n",
    "    #############################################################################################################################################################   \n",
    "    #As an output we are going to get two files: one showing all \"de novo\" proteins and their information and another showing the sum up of this information. \n",
    "    #Archivo con todos los genes de novo, Orthogrupos y nodos\n",
    "    #Creating a tsv file to save all the duplications of each node\n",
    "    De_novo_file = open(str(date.today()) + '-' + Output + \"_De_novo_Analysis.tsv\", \"w\")\n",
    "\n",
    "    writer = csv.writer(De_novo_file, delimiter = '\\t')\n",
    "    writer.writerow(['Genes','Orthogroup', 'Clade', 'Node'])\n",
    "    for key, value in De_novo_gene_information.items():\n",
    "        writer.writerow([key, value[0], value[1], value[2]])\n",
    "\n",
    "    De_novo_file.close()\n",
    "\n",
    "    #############################################################################################################################################################\n",
    "    #File containing the overall 'de novo' information\n",
    "    De_novo_Overall = open(str(date.today()) + '-' + Output + \"_De_novo_Overall.tsv\", \"w\")\n",
    "\n",
    "    writer = csv.writer(De_novo_Overall, delimiter = '\\t')\n",
    "    writer.writerow(['Node','Specie','De_novo_genes', 'De_novo_events', 'Branch_length', 'Normalized_events'])\n",
    "\n",
    "    #Esta i es para coger el número de count_Orthogroups que toca\n",
    "    i = 0\n",
    "    for specie in Clade_Path:\n",
    "        #El counter te da el número de genes de novo\n",
    "        counter = 0 \n",
    "        for element in De_novo_gene_information:\n",
    "            if De_novo_gene_information[element][1] == specie:\n",
    "                counter += 1  \n",
    "        if Specie_Name in Species_Path[i]:\n",
    "            writer.writerow([specie,specie,counter,count_Orthogroups[i],Branch_Length.get(Clade_Path.get(Species_Path[i]) + Added), count_Orthogroups[i]/(Branch_Length.get(Clade_Path.get(Species_Path[i]) + Added)*1000)])\n",
    "        if Branch_Length.get(Clade_Path.get(Species_Path[i])) is None and i != 0:\n",
    "            writer.writerow([Tree_Path[i-1],specie,counter,count_Orthogroups[i],0,0])\n",
    "        elif Branch_Length.get(Clade_Path.get(Species_Path[i])) is not None:\n",
    "            writer.writerow([Tree_Path[i-1],specie,counter,count_Orthogroups[i],Branch_Length.get(Clade_Path.get(Species_Path[i])),count_Orthogroups[i]/(Branch_Length.get(Clade_Path.get(Species_Path[i]))*1000)])\n",
    "        i += 1\n",
    "    De_novo_Overall.close()    \n",
    "\n",
    "    print('Analysis done, new file: ', str(date.today()) + '-' + Output + \"_De_novo_Analysis.tsv\")\n",
    "    print('Analysis done, new file: ', str(date.today()) + '-' + Output + \"_De_novo_Overall.tsv\")\n",
    "\n",
    "else:\n",
    "    Configurations = ['If you are not sure about how to use this settings.txt file, please read the README.md','Specie_Name:','Reference_Specie:','OrthoFinder_path:','Reference_Specie_fasta_path:','Output:','Support:']\n",
    "    with open ('Gene-B_Phylo_settings.txt','w') as file:\n",
    "        file.writelines(\"%s\\n\" % i for i in Configurations)\n",
    "    print('settings.txt file created')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
